### PR TITLE
[master < E0108-MG] Websocket support

### PR DESCRIPTION
**Pull request by antonio2368**
[master < Epic] PR
- [x] Check, and update documentation if necessary
- [x] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [x] Write E2E tests
- [x] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch

### CODE REVIEW

File: src/auth/models.cpp [hunk 1]
```diff
@@ -30,6 +30,14 @@ DEFINE_string(auth_password_strength_regex, default_password_regex.data(),
...
 namespace auth {
 
+// Constant list of all available permissions.
+constexpr std::array kPermissionsAll = {
+    Permission::MATCH,      Permission::CREATE,    Permission::MERGE,       Permission::DELETE,
+    Permission::SET,        Permission::REMOVE,    Permission::INDEX,       Permission::STATS,
+    Permission::CONSTRAINT, Permission::DUMP,      Permission::AUTH,        Permission::REPLICATION,
+    Permission::DURABILITY, Permission::READ_FILE, Permission::FREE_MEMORY, Permission::TRIGGER,
+    Permission::CONFIG,     Permission::STREAM,    Permission::WEBSOCKET};
+
```
Comment (antonio2368) [inline src/auth/models.cpp:40]: delete this

File: src/query/frontend/ast/ast.lcp [hunk 2]
```diff
@@ -2288,7 +2289,11 @@ const std::vector<AuthQuery::Privilege> kPrivilegesAll = {
     AuthQuery::Privilege::DURABILITY,
     AuthQuery::Privilege::FREE_MEMORY, AuthQuery::Privilege::TRIGGER,
     AuthQuery::Privilege::CONFIG,      AuthQuery::Privilege::STREAM,
+<<<<<<< HEAD
     AuthQuery::Privilege::MODULE_READ, AuthQuery::Privilege::MODULE_WRITE};
+=======
+    AuthQuery::Privilege::WEBSOCKET};
+>>>>>>> e15495b7 (Add websocket authentication (#322))
```
Comment (antonio2368) [inline src/query/frontend/ast/ast.lcp:2296]: modify this

File: tests/unit/cypher_main_visitor.cpp [hunk 3]
```diff
@@ -2172,6 +2172,8 @@ TEST_P(CypherMainVisitorTest, GrantPrivilege) {
                    {AuthQuery::Privilege::CONFIG});
   check_auth_query(&ast_generator, "GRANT STREAM TO user", AuthQuery::Action::GRANT_PRIVILEGE, "", "", "user", {},
                    {AuthQuery::Privilege::STREAM});
+  check_auth_query(&ast_generator, "GRANT WEBSOCKET TO user", AuthQuery::Action::GRANT_PRIVILEGE, "", "", "user", {},
```
Comment (antonio2368) [inline tests/unit/cypher_main_visitor.cpp:2175]: Add for MODULE_READ/WRITE

File: src/communication/websocket/session.hpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  bool IsConnected() const;
+
+ private:
+  using PlainWebSocket = boost::beast::websocket::stream<boost::beast::tcp_stream>;
+  using SSLWebSocket = boost::beast::websocket::stream<boost::beast::ssl_stream<boost::beast::tcp_stream>>;
+
+  explicit Session(tcp::socket &&socket, ServerContext &context, AuthenticationInterface &auth);
+
+  void DoWrite();
+  void OnWrite(boost::beast::error_code ec, size_t bytest_transferred);
```
Comment (jbajic) [inline src/communication/websocket/session.hpp:56]: Fix typo bytest

File: src/auth/models.hpp [hunk 5]
```diff
@@ -48,7 +49,8 @@ const std::vector<Permission> kPermissionsAll = {
     Permission::SET,        Permission::REMOVE,    Permission::INDEX,       Permission::STATS,
     Permission::CONSTRAINT, Permission::DUMP,      Permission::AUTH,        Permission::REPLICATION,
     Permission::DURABILITY, Permission::READ_FILE, Permission::FREE_MEMORY, Permission::TRIGGER,
-    Permission::CONFIG,     Permission::STREAM,    Permission::MODULE_READ, Permission::MODULE_WRITE};
+    Permission::CONFIG,     Permission::STREAM,    Permission::MODULE_READ, Permission::MODULE_WRITE,
+    Permission::WEBSOCKET};
```
Comment (antaljanosbenjamin) [inline src/auth/models.hpp:53]: This should be in the source file.

File: src/communication/websocket/auth.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+// licenses/APL.txt.
+
+#include "communication/websocket/auth.hpp"
+
+#include <string>
+
+namespace communication::websocket {
+
+bool SafeAuth::Authenticate(const std::string &username, const std::string &password) const {
+  // TODO: Make ReadLock after dealing with Authenticate
```
Comment (antaljanosbenjamin) [inline src/communication/websocket/auth.cpp:19]:
```suggestion
```

File: src/communication/websocket/auth.cpp [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+bool SafeAuth::HasUserPermission(const std::string &username, const auth::Permission permission) const {
+  if (const auto user = auth_->ReadLock()->GetUser(username); user) {
+    return user->GetPermissions().Has(permission) == auth::PermissionLevel::GRANT;
+  }
+  return false;
+}
+
+bool SafeAuth::HasAnyUsers() const { return auth_->ReadLock()->HasUsers(); }
+}  // namespace communication::websocket
```
Comment (antaljanosbenjamin) [inline src/communication/websocket/auth.cpp:31]:
```suggestion
}  // namespace communication::websocket

```

File: src/communication/websocket/auth.hpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  bool Authenticate(const std::string &username, const std::string &password) const override;
+
+  bool HasUserPermission(const std::string &username, auth::Permission permission) const override;
+
+  bool HasAnyUsers() const override;
+
+ private:
+  utils::Synchronized<auth::Auth, utils::WritePrioritizedRWLock> *auth_;
+};
+}  // namespace communication::websocket
```
Comment (antaljanosbenjamin) [inline src/communication/websocket/auth.hpp:44]:
```suggestion
}  // namespace communication::websocket

```

File: src/memgraph.cpp [hunk 9]
```diff
@@ -29,8 +29,12 @@
 #include <gflags/gflags.h>
 #include <spdlog/common.h>
 #include <spdlog/sinks/daily_file_sink.h>
+#include <spdlog/sinks/dist_sink.h>
 #include <spdlog/sinks/stdout_color_sinks.h>
 
+#include "communication/websocket/auth.hpp"
+#include "communication/websocket/server.hpp"
+
 #include "communication/bolt/v1/constants.hpp"
```
Comment (antaljanosbenjamin) [inline src/memgraph.cpp:38]:
```suggestion
#include "communication/bolt/v1/constants.hpp"
#include "communication/websocket/auth.hpp"
#include "communication/websocket/server.hpp"
```

File: src/memgraph.cpp [hunk 10]
```diff
@@ -43,6 +47,7 @@
 #include "query/procedure/module.hpp"
 #include "query/procedure/py_module.hpp"
 #include "requests/requests.hpp"
+#include "spdlog/spdlog.h"
```
Comment (antaljanosbenjamin) [inline src/memgraph.cpp:50]:
```suggestion
```

If necessary, then it should go the above block.

**Review by antaljanosbenjamin — APPROVED** [review]

File: tests/e2e/websocket/websocket.cpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
```
Comment (jbajic) [inline tests/e2e/websocket/websocket.cpp:10]: Add newline

File: tests/e2e/websocket/workloads.yaml [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+workloads:
+  - name: "Websocket"
+    binary: "tests/e2e/websocket/memgraph__e2e__websocket"
+    args: ["--bolt-port", *bolt_port, "--monitoring-port", *monitoring_port]
+    <<: *template_cluster
+  - name: "Websocket SSL"
+    binary: "tests/e2e/websocket/memgraph__e2e__websocket_ssl"
+    args: ["--bolt-port", *bolt_port, "--monitoring-port", *monitoring_port]
+    <<: *template_cluster_ssl
```
Comment (jbajic) [inline tests/e2e/websocket/workloads.yaml:37]: Add newline

File: src/communication/context.cpp [hunk 13]
```diff
@@ -110,43 +108,47 @@ ServerContext::ServerContext(const std::string &key_file, const std::string &cer
...
+  return ctx_;
+}
+SSL_CTX *ServerContext::context_clone() {
+  MG_ASSERT(ctx_);
+  SSL_CTX_up_ref(ctx_);
+  return ctx_;
+}
 
-bool ServerContext::use_ssl() { return use_ssl_; }
+bool ServerContext::use_ssl() { return ctx_ != nullptr; }
```
Comment (kostasrim) [inline src/communication/context.cpp:152]:
```suggestion
bool ServerContext::use_ssl() const { return ctx_ != nullptr; }
```

File: src/communication/context.cpp [hunk 14]
```diff
@@ -110,43 +108,47 @@ ServerContext::ServerContext(const std::string &key_file, const std::string &cer
...
       SSL_CTX_set_verify(ctx_, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, nullptr);
     }
   }
 }
 
-ServerContext::ServerContext(ServerContext &&other) noexcept : use_ssl_(other.use_ssl_), ctx_(other.ctx_) {
-  other.use_ssl_ = false;
-  other.ctx_ = nullptr;
-}
+ServerContext::ServerContext(ServerContext &&other) noexcept : ctx_(other.ctx_) { other.ctx_ = nullptr; }
```
Comment (kostasrim) [inline src/communication/context.cpp:117]:
```suggestion
ServerContext::ServerContext(ServerContext &&other) noexcept {
 std::swap(ctx_, other.ctx);
}
```

File: src/communication/websocket/session.cpp [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+    return;
+  }
+
+  if (!IsAuthenticated()) {
+    auto response = nlohmann::json();
+    auto auth_failed = [this, &response](const std::string &message) {
+      response["success"] = false;
+      response["message"] = message;
+      MG_ASSERT(messages_.empty());
+      messages_.push_back(make_shared<std::string>(response.dump()));
```
Comment (kostasrim) [inline src/communication/websocket/session.cpp:184]:
```suggestion
      messages_.push_back(std::make_shared<std::string>(response.dump()));
```

File: src/communication/websocket/session.cpp [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+
+      if (const auto result = Authorize(creds); result.HasError()) {
+        std::invoke(auth_failed, result.GetError());
+        return;
+      }
+      response["success"] = true;
+      response["message"] = "User has been successfully authenticated!";
+      MG_ASSERT(messages_.empty());
+      authenticated_ = true;
+      messages_.push_back(make_shared<std::string>(response.dump()));
```
Comment (kostasrim) [inline src/communication/websocket/session.cpp:200]:
```suggestion
      messages_.push_back(std::make_shared<std::string>(response.dump()));
```

**Pull request merged**